tsnet/tsnet.go

```diff
@@ -169,9 +169,38 @@ func (s *Server) Dial(ctx context.Context, network, address string) (net.Conn, e
 	if err := s.Start(); err != nil {
 		return nil, err
 	}
+	if err := s.awaitRunning(ctx); err != nil {
+		return nil, err
+	}
 	return s.dialer.UserDial(ctx, network, address)
 }
 
+// awaitRunning waits until the backend is in state Running.
+// If the backend is in state Starting, it blocks until it reaches
+// a terminal state (such as Stopped, NeedsMachineAuth)
+// or the context expires.
+func (s *Server) awaitRunning(ctx context.Context) error {
+	st := s.lb.State()
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		switch st {
+		case ipn.Running:
+			return nil
+		case ipn.Stopped, ipn.NeedsMachineAuth:
+			return fmt.Errorf("tsnet: backend in state %v", st)
+		default: // Starting, NeedsLogin, ...
+			s.lb.WatchNotifications(ctx, ipn.NotifyInitialState, nil, func(n *ipn.Notify) (keepGoing bool) {
+				if n.State != nil {
+					st = *n.State
+				}
+				return st == ipn.Starting
```

Comment:
Do you also want to break on `ipn.Stopped` / `ipn.NeedsMachineAuth` here so that you can return an error?

Reply:
It does break on those.... if it's not Starting exactly, we return false and stops the watch loop.